signals/src/main/java/com/vaadin/signals/impl/MutableTreeRevision.java

```diff
@@ -0,0 +1,857 @@
+            } else if (updatedNodes.containsKey(id)) {
+                return Optional.ofNullable((Data) updatedNodes.get(id));
+            } else {
+                return MutableTreeRevision.this.data(id);
+            }
+        }
+
+        private void useData(Id nodeId, BiConsumer<Data, Id> consumer) {
+            assert result == null;
+
+            Id id = resolveAlias(nodeId);
+            data(id).ifPresentOrElse(node -> consumer.accept(node, id), () -> {
+                fail("Node not found");
+            });
+        }
+
+        private void updateData(Id nodeId, UnaryOperator<Data> updater) {
+            useData(nodeId, (node, id) -> {
+                Data updatedNode = updater.apply(node);
+                if (updatedNode != node) {
+                    updatedNodes.put(id, updatedNode);
+                }
+            });
+        }
+
+        private JsonNode value(Id nodeId) {
+            return data(nodeId).map(Data::value).orElse(null);
+        }
+
+        private void setValue(Id nodeId, JsonNode value) {
+            updateData(nodeId,
+                    node -> new Data(node.parent(), command.commandId(),
+                            node.scopeOwner(), value, node.listChildren(),
+                            node.mapChildren()));
+        }
+
+        private Optional<List<Id>> listChildren(Id parentId) {
+            return data(parentId).map(Data::listChildren);
+        }
+
+        private boolean isChildAt(Id parentId, int index, Id expectedChild) {
+            assert expectedChild != null;
+
+            if (index < 0) {
+                return false;
+            }
+
+            Id idAtIndex = listChildren(parentId).map(children -> {
+                if (index >= children.size()) {
+                    return null;
+                }
+                return children.get(index);
+            }).orElse(null);
+
+            return isSameNode(idAtIndex, expectedChild);
+        }
+
+        private Optional<Id> mapChild(Id nodeId, String key) {
+            return data(nodeId).map(Data::mapChildren)
+                    .map(children -> children.get(key));
+        }
+
+        private boolean isSameNode(Id a, Id b) {
+            return Objects.equals(resolveAlias(a), resolveAlias(b));
+        }
+
+        private boolean detach(Id nodeId) {
+            useData(nodeId, (node, id) -> {
+                if (id.equals(Id.ZERO)) {
+                    fail("Cannot detach the root");
+                    return;
+                }
+
+                Id parentId = node.parent();
+                if (parentId == null) {
+                    fail("Node is not attached");
+                    return;
+                }
+
+                Data parentData = data(parentId).get();
+
+                String key = parentData.mapChildren().entrySet().stream()
+                        .filter(entry -> entry.getValue().equals(id))
+                        .map(Entry::getKey).findAny().orElse(null);
```

Comment:
Nit: Prioritize `findAny()` before mapping:
```suggestion
                String key = parentData.mapChildren().entrySet().stream()
                        .filter(entry -> entry.getValue().equals(id))
                        .findAny().map(Entry::getKey).orElse(null);
```